openedx/openedx-platform
# [ENT-334] Add client-side registration form validation.

## Front-end

### Focus on a field (input/textarea/select ONLY)

![selection_003](https://user-images.githubusercontent.com/10018065/28566942-28361286-714a-11e7-9b79-5a05e5b0983d.png)

### Blur and get a SUCCESS validation (input/textarea/select)

![selection_002](https://user-images.githubusercontent.com/10018065/28566789-a1505998-7149-11e7-83e2-b6566fe18bce.png)

### Check and get a SUCCESS validation (checkbox)

![selection_006](https://user-images.githubusercontent.com/10018065/28566808-b5602b98-7149-11e7-948b-820d227c04c2.png)

### Focus on SUCCESS validation (input/textarea/select ONLY)

![selection_004](https://user-images.githubusercontent.com/10018065/28566922-101466ee-714a-11e7-84b4-8f8adf6cef38.png)

### SUCCESS validation fades (after 3 seconds) (input/textarea/select)

![selection_005](https://user-images.githubusercontent.com/10018065/28566876-e4c0b75e-7149-11e7-9088-6535d9aca2ce.png)

### SUCCESS validation fades (after 3 seconds) (checkbox)

![selection_007](https://user-images.githubusercontent.com/10018065/28566859-d88e464a-7149-11e7-9664-22da328495a2.png)

### Blur and get an ERROR validation (input/textarea/select)

![selection_001](https://user-images.githubusercontent.com/10018065/28566693-5e8e3a76-7149-11e7-87d0-c08aacbc32c7.png)

### Uncheck and get an ERROR validation (checkbox)

![image](https://user-images.githubusercontent.com/10018065/28566671-506a3896-7149-11e7-8aef-274e826d56a1.png)


## Back-end
This implements a validation API for registration, where a client makes AJAX calls to the endpoints at

```bash
/api/user/v1/validation/registration
```

requesting validation decisions on each input. Inputs are of the form

```json
{
    "name": "Uman Shahzad",
    "email": "edx@email.com",
    "confirm_email": "edx@email.com",
    "username": "myusername",
    "password": "mypassword",
    "country": "PK"
}
```

However, a user may just as well input

```json
{
    "email": "edx@email.com"
}
```

to get a validation decision for just the `email` field alone. The same applies for all other form fields for which there exists a handler (see below for more info on handlers).

Responses come in JSON format and provide strings for each input form field dependent on the type of validation error; if no error, then empty string to indicate OK. For example, in response to the latter input above, which is a valid email:

```json
{
    "validation_decisions": {
        "email": ""
    }
}
```

If the email had any issues in its format, we'd get an appropriate error message response in place of `""`. Clients can then use this error message directly on the interface with styling.

See documentation for more detail on what we can input & get back.

### Handlers

A handler in this API's architecture is just a method. These handlers simply return an error message string, or an empty string if no error occurred in validation. There currently exist implementations of handlers for 

- `name`
- `email`
- `confirm_email`
- `password`
- `username`
- `country`

Extending the API for more handlers is as simple as adding a method for handling, that takes in a `request` and outputs an error message (or an empty string), then adding this method to the dictionary of available validation handlers.

## Other

**JIRA tickets**: Implements ENT-334's back-end version (front-end to come).

**Discussions**: The architecture of the validation API is covered extensively in the `RegistrationValidationView`'s documentation, and above.

**Dependencies**: None

**Screenshots**: Back-end only right now -- screenshots to come when front-end work is up (next week).

**Sandbox URL**: https://pr15375.sandbox.opencraft.hosting/

**Deployment targets**: edx.org and edge.edx.org

**Merge deadline**: Hopefully by start of July.

**Testing instructions**:

1. Go to the registration page with this branch running on your devstack or the sandbox (see if URL is available above).
1. Input into any field proper and improper inputs for that field, click somewhere else or tab, and see if it appropriately notifies you right away that there is an issue or that you entered correctly.

**Author notes and concerns**:

1. The `RegistrationValidationView`'s handlers could use some refactoring, potentially -- instead of individual methods, we can make these instances of classes like `PasswordValidation`, etc., but I believe this may complicate things unnecessarily because the method-handler point of view is in itself still perfectly fine.
2. The cross/check FA icons are visualized on the requirements as being on the right side of the input field. However, this could require a large amount of CSS code to change, because of the way the registration container is made, and potentially break mobile or simply not work on that platform either way (because input fields take up full horizontal space on small devices). I propose a different way of handling the icons:

![](http://i.imgur.com/6nb2BYm.png)
and
![](http://i.imgur.com/lhZGGWh.png)

Having them right next to the label is clear, obvious, viewable on mobile without mobile-specific changes, and a cake to code up (I already did). It would also be possible to do this cleanly for other fields which don't take up full horizontal space (like dropdown lists).

**Reviewers**
- [ ] @brittneyexline @douglashall @chris-mike @cptvitamin 

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
EDXAPP_REGISTRATION_EXTRA_FIELDS:
  confirm_email: "required"
  level_of_education: "required"
  gender: "required"
  year_of_birth: "required"
  mailing_address: "required"
  goals: "required"
  honor_code: "required"
  terms_of_service: "required"
  city: "required"
  country: "required"
```
